### PR TITLE
msg/async/rdma: Move resource handling to Device

### PR DESCRIPTION
**Pull request by Adirl**

### CODE REVIEW

**Comment by Adirl** [comment]
@yuyuyu101 @orendu @amirv @saritz @DanielBar-On

**Comment by Adirl** [comment]
@yuyuyu101 
please take a look when you can 
thanks

File: src/msg/async/rdma/Infiniband.cc [hunk 1]
```diff
@@ -890,3 +768,33 @@ const char* Infiniband::qp_state_string(int status) {
     default: return " out of range.";
   }
 }
+
+void Infiniband::handle_pre_fork()
+{
+  device->uninit();
```
Comment (yuyuyu101) [inline src/msg/async/rdma/Infiniband.cc]: if multi device, I think we need to make all devices uninit?

**Comment by yuyuyu101** [comment]
good cleanup, we add multi device supports here. but we only make use of one device now even after this pr, right?

**Comment by yuyuyu101** [comment]
if multi device, I think infiniband only need to operate DeviceList instead of one device.

File: src/msg/async/rdma/Infiniband.cc (hunk 1 above)
Comment (amirv) [inline src/msg/async/rdma/Infiniband.cc]: Correct. This commit is a preparation only. It only initializes the selected device (from the config parameter).
I will soon send you a patch that enable using more than 1 device, and the uninit() will be modified accordingly.

**Comment by amirv** [comment]
Yes, it is my mistake that I didn't share with you the patch plan:

This is the first part of 3 in the series. Every such part consist of few patches to make it easier for review.

In this part there are 2 patches which are only cleanup and code rearranging without changing the actual logic. Still only one device is being supported.

In part 2 we will send some patches that finishes the multi device support, and make the init, uninit and polling multidevice aware. Once this accepted, everything is ready to the last part.

In the last part we will add the RDMA-CM basic support. Which is our target for the end of the month.
A config param will enable the user to select on runtime if he wants RDMA-CM or TCP connection establishment.

Thanks,
Amir

**Comment by yuyuyu101** [comment]
@Adirl do we passed the basic tests?

**Comment by Adirl** [comment]
@yuyuyu101 yes we did

**Comment by Adirl** [comment]
@yuyuyu101 
anything else needed here? 

Thanks

**Pull request merged**